impeller/display_list/dl_dispatcher.cc

```diff
@@ -722,23 +722,10 @@ void DlDispatcherBase::saveLayer(const SkRect& bounds,
   auto promise = options.content_is_clipped()
                      ? ContentBoundsPromise::kMayClipContents
                      : ContentBoundsPromise::kContainsContents;
-  std::optional<Rect> impeller_bounds;
-  if (!options.content_is_unbounded() || options.bounds_from_caller()) {
-    impeller_bounds = skia_conversions::ToRect(bounds);
-  }
-
-  // Empty bounds on a save layer that contains a BDF or destructive blend
-  // should be treated as unbounded. All other empty bounds can be skipped.
-  if (impeller_bounds.has_value() && impeller_bounds->IsEmpty() &&
-      (backdrop != nullptr ||
-       Entity::IsBlendModeDestructive(paint.blend_mode))) {
-    impeller_bounds = std::nullopt;
-  }
-
+  std::optional<Rect> impeller_bounds = skia_conversions::ToRect(bounds);
   GetCanvas().SaveLayer(paint, impeller_bounds, ToImageFilter(backdrop),
                         promise, total_content_depth,
-                        options.can_distribute_opacity(),
-                        options.bounds_from_caller());
+                        options.can_distribute_opacity());
```

Comment:
I think you also need `options.content_is_unbounded()`.

- The input will be unbounded if this SL has a BDF.
- The input ?might? also be unbounded if this SL is applied with a flooding CF (I'm still wrapping my brain around that case).
- But, the input will also be unbounded if any child content between SL and Restore is itself output-unbounded. You can't tell this case from the paint object or the presence of a BDF, you have to look at the flags in the options.

The supplied bounds are still provided because the flag overrides them and we did calculate them, so why not, but they don't promise that the content of this SL is bounded if that flag is present.

If there is unbounded child content, DL might substitute the clip at the moment that content is discovered - but only if that enclosing clip is inside the same DL/layer. If there has not yet been a restricting clip before we get to the unbounded content inside the SL, then the SL itself is marked unbounded via the options.

So, there are 3 conditions that lead to unbounded contents.

Reply:
The unbounded content should be handled in the line about where I replace the content bounds with std::nullopt:

https://github.com/flutter/engine/blob/main/impeller/display_list/dl_dispatcher.cc#L726-L728

this doesn't solve the problem of impeller/dl speaking different terms of unbounded. Longer term, we should probably just plumb this flag through.

Reply:
> The input ?might? also be unbounded if this SL is applied with a flooding CF (I'm still wrapping my brain around that case).

We handle that by only applying clear color on srcOver, but layer on that is the flood clip check in entity_pass

Reply:
The unbounded line was deleted. You are pointing to what the code "used to do"...

Reply:
🤦

Reply:
I re-unbounded it.

Reply:
I have to think more about that. The unbounded is a reason to ignore the bounds, but I'm not sure that user supplied bounds are a reason to use them even in the unbounded case.

Scratches head...

Reply:
Oh right, we discussed this! That is why I deleted the line, but I need to add it back.

Reply:
OK, removed bounds from caller. We just check if its unbounded.

Reply:
agh, I think we hit this already. If a saveLayer contains unbounded content, but does not have a bdf or destructive blend, but the saveLayer was bounded - then bounds.is_clipped should be true and we don't null out the bds.

Reply:
Okay so this is a case where maybe there is actually a DL misunderstanding. On the test ImageFilteredSaveLayerWithUnboundedContents we expect the content to be clipped via the saveLayer bounds. This saveLayer has a specified bounds of (0, 0, 200, 200) but `options.content_is_unbounded` is true.

I guess, its technically correct that this saveLayer, which contains a drawPaint, is unbounded - but we decided to treat is bounded so maybe content_is_unbounded should return false?